### ISSUE TITLE
Numeric coercions

### ISSUE THREAD

**Issue by owickstrom** [oden-lang/oden#32]
Coercion should be explicit, like in Go (https://tour.golang.org/basics/13). Consider introducing special AST nodes in Core for coercion to make it more clear in the inference process.

**Comment by owickstrom** [comment]
This might prove a bit tricky when it comes to inference. For example, what should the type of `x` be inferred to in the following code?

```go
f(x) -> int32(x)
```

This is where I previously thought about type classes, but I'm not sure it's a good fit for Oden. I have very early thoughts about some other kind of constraint system that can be used for expressing this type. It ties into some design ideas about how interfaces should work as well *(spoiler: row polymorphism, like records, but for interface method sets)*.

**Comment by Heimdell** [comment]
```haskell
f :: forall a. Coersible a int32 => a -> int32
```
?

**Comment by owickstrom** [comment]
@Heimdell yeah, that's the type class approach I'm working on right now. As they will be a bit different from Haskell type classes and from Go interfaces I've began referring to them as _Protocols_. You can look at the preliminary sketches at https://gist.github.com/owickstrom/abbdf5ad81b317064c63d9df4badb8c0. As you can see they're always qualified by the protocol name, but otherwise called like regular functions (not like receiver functions in Go). The goal is to be able to translate automatically between protocols and existing interfaces in Go.

In the case of numeric conversions it could be something like this:

```go
f(x) -> ToInt::int32(x)
```

... and the Go backend would have a "special" instance for that protocol using Go's type coercions.

**Comment by Heimdell** [comment]
This looks ok, but why the protocol's methods are fully qualified? You can distinguish normal functions and methods by AST metadata. Or, maybe you want to have different protocols containing same methods?

By the way, do you plan supporting userspace operators? For instance, `mf <$> mx` as `__operator_gt_dollar_lt(mf, mx)`.

**Comment by owickstrom** [comment]
Yes, because of how interface method scoping in Go works it needs to be qualified if Go interfaces are going to work seamlessly with Oden. I haven't found a good way around this yet.
